### DISCUSSION TITLE
How to sync app to tenant?

### DISCUSSION THREAD

**Discussion by j2l** [lorinkoz/django-pgschemas#264] [Q&A]
I'm continuing to learn this great extension.
By mistake I first set `tenant_app` in static www, and migrated:
```
"default": {
        "TENANT_MODEL": "app.Client",
        "DOMAIN_MODEL": "app.Domain",
        "APPS": [
            'django.contrib.admin',
            "django.contrib.auth",
            "django.contrib.sessions",
            'django.contrib.messages',
            "backend",
        ],
        "URLCONF": "backend.urls",
    },
    "www": {
        "APPS": [
            "django.contrib.auth",
            "django.contrib.sessions",
            "tenant_app",
        ],
```

then I moved it to default:
```
"default": {
        "TENANT_MODEL": "app.Client",
        "DOMAIN_MODEL": "app.Domain",
        "APPS": [
            'django.contrib.admin',
            "django.contrib.auth",
            "django.contrib.sessions",
            'django.contrib.messages',
            "backend",
            "tenant_app",
        ],
        "URLCONF": "backend.urls",
    },
    "www": {
        "APPS": [
            "django.contrib.auth",
            "django.contrib.sessions",
        ],
```
After migrating, I encountered 2 problems:
- I can still access tenant_app model in `www` admin (no sync after migrating again)
- Already created tenants (client1) with already existing schema didn't get the model update (no sync after migrating again).

But, hopefully, creating a new tenant and migrating created the model in the new tenant.

**So I guess I missed a sync, or a way to delete a dynamic tenant with its schema?**
(which is mandatory for The right to be forgotten (RTBF) in Europe I think, and memory wise for Saas).
**I also wonder how to add a new app to all existing tenants.**

**Comment by lorinkoz** [comment]
I recommend you start from scratch with the right apps on the right tenants. Otherwise see https://django-pgschemas.readthedocs.io/en/stable/troubleshooting.html#moving-apps-between-schemas

**Comment by j2l** [comment]
Thank you!

> I recommend you start from scratch with the right apps on the right tenants. 

Ah, perfect world indeed! :smile_cat:  Will it ever exist?
More seriously, I'm pretty sure in less than 5 years, I'd like to add an app to an existing users base.

`manage.py migrate tenant_app zero --fake -s www` unlinked the tenant_app from the tenant but without dropping the tables, `python manage.py migrate` leading to `django.db.utils.ProgrammingError: relation "tenant_app_article" already exists`. Manually deleting the tables in the right schema fixed it.

But they are recreated at the next migration. I don't get where the info is located that `tenant_app` (default) should be applied to `www` (static).

`python manage.py migrate tenant_app -s client1` did the job to add an app to an existing tenant.

**I still don't get how to delete a dynamic tenant with its schema**

**Reply by lorinkoz** [reply]
> I still don't get how to delete a dynamic tenant with its schema

If you delete the dynamic tenant it should delete the schema as part of the process.

**Reply by j2l** [reply]
How?

**Reply by lorinkoz** [reply]
How to delete the dynamic tenant?

`Tenant.objects.filter(filter_tenant_to_delete).delete()` in a `python manage.py shell`

If you were expecting to do it via Django admin, I assume you should enable the tenant model in the admin config of, e.g. your "www" static tenant, and then you could select the tenant in the Django admin and delete it there.

**Reply by j2l** [reply]
`Client.objects.filter(name='client1').delete()` works, but schema is still present in the DB.
And even if `python manage.py migrate zero -s client1` throws `CommandError: No schema found for 'client1'`, it is still very present in the DB:
![image](https://github.com/user-attachments/assets/5f4b8428-225f-4764-8f99-34792f776e35)


> If you were expecting to do it via Django admin, I assume you should enable the tenant model in the admin config of, e.g. your "www" static tenant

Via admin, AFAIU this is impossible because the shared_app records (called `app` in my case) must be on `public`, hence my other discussion to access public admin (to manage Client and Domain tables).
If I change app/admin.py to show Client and Domain, it appears in each tenant administration panel, because it's shared among tenants.
Without app/admin.py
![image](https://github.com/user-attachments/assets/0a415646-8515-4e2d-93d0-3908de11f156)
With app/admin.py
![image](https://github.com/user-attachments/assets/168da956-2fc6-4299-9004-f48bfe2c75a4)

**Comment by j2l** [comment]
About the delete,
> schema is still present in the DB

Could you please try and confirm?

**Comment by lorinkoz** [comment]
Heya. Were you able to resolve this or same situation>

**Reply by j2l** [reply]
Thanks for the follow up, same situation here.
Any fix?
